src/Notes.cc

```diff
@@ -361,6 +363,47 @@ NotePairs::addStrList(const SBuf &key, const SBuf &values, const CharacterSet &d
     AppendTokens(entries, key, values, delimiters);
 }
 
+void
+NotePairs::importFromHelper(const SBuf &rawNotes)
+{
+    // XXX: Reduce code duplication with Notes::parse()
+    Parser::Tokenizer tok(rawNotes);
+    while (!tok.atEnd()) {
+        static const auto nameChars = (
+            CharacterSet::ALPHA +
+            CharacterSet::DIGIT +
+            CharacterSet("other", "-._")
+        ).rename("helper-note-name");
+
+        static const auto valueChars = (
+            CharacterSet::WSP +
+            CharacterSet("comma", ",") // we do not support CSVs (yet?)
+        ).complement("helper-note-value");
+
+        SBuf name;
+        if (!tok.prefix(name, nameChars))
+            throw TextException(ToSBuf("malformed from-helper annotation name near: ", tok.remaining()), Here());
+
+        if (!tok.skip('='))
+            throw TextException(ToSBuf("missing '=' after from-helper annotation near: ", tok.remaining()), Here());
+
+        SBuf value;
+        if (!tok.prefix(value, valueChars))
+            throw TextException(ToSBuf("malformed from-helper annotation value: ", tok.remaining()), Here());
+
+        const auto start = value.at(0);
+        if (start == '"' || start == '`' || start == '\'' || start == '\\')
+            throw TextException(ToSBuf("unsupported complex from-helper annotation value near: ", value), Here());
+
+        if (!tok.atEnd() && !tok.skipAll(CharacterSet::WSP))
+            throw TextException(ToSBuf("from-helper annotation missing whitespace after name=value near: ", tok.remaining()), Here());
+
+        // TODO: reject blacklisted names?
```

Comment:
I forgot what blacklisted means in this comment exactly. Our [wiki](https://wiki.squid-cache.org/Features/AddonHelpers#SSL_server_certificate_validator) does not list any reserved names. We should either be more specific or address this TODO. All other factors being equal, addressing it would be better, of course.

The overall idea here is that if we might add support for some name later, then we must prohibit it _now_, so that our future additions do not clash with what admins may start to use.

Perhaps we should ban all(?) names that do not end with an underscore (names that do are reserved for admin use)? If that is not possible for some reason, but Squid code processing certificate validator helper responses does not support any of the names checked in Notes::validateKey() (yet), then I wonder whether we should call Notes::validateKey() here?

Reply:
> ban all(?) names that do not end with an underscore

We have had several names not ending with '_', such as cert_ID, error_name_ID, etc. Also we have clt_conn_tag added in this PR. So I would not ban all such names to avoid breaking existing helpers.

>  certificate validator helper responses does not support any of the names checked in Notes::validateKey() (yet)

That is true - I think we can validateKey() there. According to the Wiki [page](https://wiki.squid-cache.org/Features/AddonHelpers), each helper uses some of names returned by validateKey() and documents others as 'reserved' (to avoid conflicts with other helpers, I guess). Following this logic, our certificate validator should mark all these names as 'reserved', since it does not use any.

Reply:
cert_ID/error_name_ID/clot_conn_tag/etc are not really in scope here because they are Squid-recognized helper response parameters (rather than transaction annotations). We will not ban them no matter what, of course. The word "all" was probably too broad. I should have said something like "all unsupported" or "all that are not already recognized by Squid", but even that can be misinterpreted. We do not have well-defined terminology here, unfortunately, but keep in mind the end goal (which I hope my comments below clarify).

> to avoid conflicts with other helpers, I guess

Primarily, we reserve a name (for possible future Squid use) to allow _adding_ new Squid-recognized names without conflicting with _old_ helpers (that might happen to use them). Imagine that we want to add support for server_conn_tag. If we do not restrict the set of names that the certificate validation helpers may emit today, then one of them (say, X) may emit server_conn_tag for its own/admin's use. When we release Squid support for an annotation named server_conn_tag, then that Squid release will silently break Squid+X because Squid+X combo was probably not using server_conn_tag exactly the same way the newly released Squid will be using it.

> our certificate validator should mark all these names as 'reserved', since it does not use any.

Yes, but this will be a very small step forward. I hope we can reserve a lot more (e.g., all that are not already recognized by Squid today) so that it is easy to add more recognized-by-Squid validation helper annotations in the future.